sdk/eventhub/azure-eventhub/azure/eventhub/aio/_transport/_pyamqp_transport_async.py

```diff
@@ -214,16 +214,26 @@ async def receive_messages_async(consumer, batch, max_batch_size, max_wait_time)
         # pylint:disable=protected-access
         consumer._callback_task_run = True
         consumer._last_callback_called_time = time.time()
-        callback_task = asyncio.ensure_future(
+        callback_task = asyncio.create_task(
             PyamqpTransportAsync._callback_task(consumer, batch, max_batch_size, max_wait_time)
         )
-        receive_task = asyncio.ensure_future(PyamqpTransportAsync._receive_task(consumer))
+        receive_task = asyncio.create_task(PyamqpTransportAsync._receive_task(consumer))
 
+        tasks = [callback_task, receive_task]
         try:
-            await receive_task
-        finally:
+            for task in asyncio.as_completed(tasks):
+                try:
+                    await task
+                    await asyncio.sleep(0)
+                except Exception: # pylint: disable=broad-except
+                    consumer._callback_task_run = False
+            for task in tasks:
+                if task.done() and task.exception():
+                    raise task.exception()
+        except asyncio.CancelledError:
             consumer._callback_task_run = False
-            await callback_task
+            await asyncio.sleep(0)
+            raise
```

Comment:
I think this can still cause asyncio warnings about cancelled pending tasks or task exceptions not retrieved.
I will use this patch in my testing this afternoon to see if I still can repro that in combination with your other fixes.

Reply:
yeah this was mostly too see if there were some other exceptions floating around, then would the main loop catch them.